librz/bin/format/elf/elf.c

```diff
@@ -417,29 +436,44 @@ static bool init_ehdr_ident(ELFOBJ *bin) {
 	return true;
 }
 
+static bool is_tiny_elf(ELFOBJ *bin) {
+	return bin->size == 45;
+}
+
+static Elf_(Half) get_tiny_elf_phnum(ELFOBJ *bin) {
+	ut8 tmp;
+	ut64 offset = 44;
+
+	if (Elf_(rz_bin_elf_read_char)(bin, &offset, &tmp)) {
+		return tmp;
+	}
+
+	return 0;
+}
+
 static bool init_ehdr_other(ELFOBJ *bin) {
-	size_t i = EI_NIDENT;
-	ut8 ehdr[sizeof(Elf_(Ehdr))] = { 0 };
-	rz_buf_read_at(bin->b, 0, ehdr, sizeof(ehdr));
-
-	if (bin->size < 32) { // tinyelf != sizeof (Elf_(Ehdr)))
-		bprintf("read (ehdr)\n");
-		return false;
-	}
-
-	bin->ehdr.e_type = READ16(ehdr, i);
-	bin->ehdr.e_machine = READ16(ehdr, i);
-	bin->ehdr.e_version = READ32(ehdr, i);
-	bin->ehdr.e_entry = RZ_BIN_ELF_READWORD(ehdr, i);
-	bin->ehdr.e_phoff = RZ_BIN_ELF_READWORD(ehdr, i);
-	bin->ehdr.e_shoff = RZ_BIN_ELF_READWORD(ehdr, i);
-	bin->ehdr.e_flags = READ32(ehdr, i);
-	bin->ehdr.e_ehsize = READ16(ehdr, i);
-	bin->ehdr.e_phentsize = READ16(ehdr, i);
-	bin->ehdr.e_phnum = READ16(ehdr, i);
-	bin->ehdr.e_shentsize = READ16(ehdr, i);
-	bin->ehdr.e_shnum = READ16(ehdr, i);
-	bin->ehdr.e_shstrndx = READ16(ehdr, i);
+	ut64 offset = EI_NIDENT;
+
+	if (!Elf_(rz_bin_elf_read_half)(bin, &offset, &bin->ehdr.e_type) ||
```

Comment:
Note, that the previous code was more efficient since it was reading all of the header in the intermediate buffer, and then reading data from the buffer, since going all over the RzBin interface could be a bit slower - and if you have many read operations it multiplies. On the other hand, the RzBin has been changed significantly since by @thestr4ng3r, so it's unclear. I recommend you to compare the performance of your solution vs the old one, and if the new is slower - use the intermediate buffer as well.

Reply:
Before

```
( repeat 10; do; rz-test db/formats/elf; done; )  105.47s user 11.56s system 373% cpu 31.366 total
```

After

```
( repeat 10; do; rz-test db/formats/elf; done; )  99.84s user 11.44s system 359% cpu 30.828 total
```

There is a slow down, but compare to the number of binary executed around 1480...

I didn't use a static buffer, because it is prone to bug.

Reply:
I would expect lto to take care of most of this when enabled.

Reply:
What do you mean by "when enabled"?